shell/platform/android/io/flutter/embedding/engine/FlutterEngine.java

```diff
@@ -200,21 +225,81 @@ public FlutterEngine(
       @NonNull PlatformViewsController platformViewsController,
       @Nullable String[] dartVmArgs,
       boolean automaticallyRegisterPlugins) {
+    doInit(
+        context,
+        flutterLoader,
+        flutterJNI,
+        platformViewsController,
+        dartVmArgs,
+        automaticallyRegisterPlugins,
+        null);
+  }
+
+  public FlutterEngine() {}
+
+  public void initAsync(
+      @NonNull Context appContext, @NonNull EngineLifecycleListener asyncInitCallback) {
+    if (asyncInitCallback == null) {
+      Log.w(
+          TAG,
+          "initAsync: callback is null, you should care lifecycle, and called other api after initCallback called");
+      asyncInitCallback =
+          new EngineLifecycleListener() {
+            @Override
+            public void onPreEngineRestart() {}
+
+            @Override
+            public void onAsyncAttachEnd(boolean success) {}
+
+            @Override
+            public void onAsyncCreateEngineEnd(boolean success) {}
+          };
+    }
+    doInit(
+        appContext,
+        FlutterLoader.getInstance(),
+        new FlutterJNI(),
+        new PlatformViewsController(),
+        null,
+        true,
+        asyncInitCallback);
+  }
+
+  private void doInit(
+      @NonNull Context context,
+      @NonNull FlutterLoader flutterLoader,
+      @NonNull FlutterJNI flutterJNI,
+      @NonNull PlatformViewsController platformViewsController,
+      @Nullable String[] dartVmArgs,
+      boolean automaticallyRegisterPlugins,
+      EngineLifecycleListener asyncInitListener) {
     this.flutterJNI = flutterJNI;
+    this.context = context;
+    this.automaticallyRegisterPlugins = automaticallyRegisterPlugins;
+    this.flutterLoader = flutterLoader;
+    this.platformViewsController = platformViewsController;
     flutterLoader.startInitialization(context.getApplicationContext());
     flutterLoader.ensureInitializationComplete(context, dartVmArgs);
 
     flutterJNI.addEngineLifecycleListener(engineLifecycleListener);
-    attachToJni();
+    if (asyncInitListener != null) {
```

Comment:
Since this is all about doing FlutterJNI.attachToNative() concurrently, you can just call that directly and not mix it into the state of the FlutterEngine instance. Do it outside this class.